fs/operations/operations.go

```diff
@@ -343,6 +344,112 @@ func CommonHash(fa, fb fs.Info) (hash.Type, *fs.HashesOption) {
 // It returns the destination object if possible.  Note that this may
 // be nil.
 func Copy(ctx context.Context, f fs.Fs, dst fs.Object, remote string, src fs.Object) (newDst fs.Object, err error) {
+	resumeOpt := &fs.OptionResume{}
+	if doResume := f.Features().Resume; doResume != nil {
+		// Creates an OptionResume that will be passed to Put/Upload
+		resumeOpt = &fs.OptionResume{ID: "", Pos: 0, SetID:
+		// SetID will be called by backend's Put/Update function if the object's upload
+		// could be resumed upon failure
+		//
+		// SetID takes the passed resume ID and Fingerprint of the object and stores it in
+		// --cache-dir so that future Copy operations can resume the upload if it fails
+		func(ID string) (err error) {
+			rootCacheDir := filepath.Join(config.CacheDir, "resume")
+			if filepath.Ext(rootCacheDir) != "" {
+				rootCacheDir = filepath.Dir(rootCacheDir)
+			}
+			// Each remote will have its own dir for cached resume files
+			dirPath := filepath.Join(rootCacheDir, f.Name(), f.Root())
+			err = os.MkdirAll(dirPath, os.ModePerm)
+			if err != nil {
+				return errors.Wrapf(err, "failed to create cache directory %v", dirPath)
+			}
+			cachePath := filepath.Join(dirPath, remote)
+			cacheFile, err := os.Create(cachePath)
+			if err != nil {
+				return errors.Wrapf(err, "failed to create cache file %v", cachePath)
+			}
+			defer func() {
+				_ = cacheFile.Close()
+			}()
+			// Check how many resume cache files already exist and remove oldest if more than limit
```

Comment:
Do we want to do this on *every* file we upload?

Reply:
compare e.g. with _curl_ tool, which activates resume logic only when user adds _--continue-at_ on command line (https://curl.se/docs/manpage.html, google "curl continue")

Reply:
I made a slight tweak so that the cache is only cleaned once per file uploaded. Previously, if a backend had to repeatedly update the cache for a single file (e.g. Chunker updating the cache with each successfully uploaded chunk) the cache was being unnecessarily cleaned during each update.

@ncw Do you have any suggestions for how we might further mitigate how often the cache is cleaned?

@ivandeex This is certainly something I can look into adding. Would we want such an option to only control whether or not to attempt a resume on an interrupted upload, or would we also want to only conditionally write the necessary information to the cache to make a later resume possible? Either with an option or maybe only for files above a certain size where resumes are more useful?

Reply:
`--resume-larger` satisfies the request.

> Would we want such an option to only control whether or not to attempt a resume on an interrupted upload, or would we also want to only conditionally write the necessary information to the cache to make a later resume possible?

I think this bit will become clear when you proceed to adding unit tests (resume normally or resume after corrupted source/destination) and making them pass.

Reply:
Once per file seems OK thinking some more about it since we are only going to activate this for large files probably which take lots of time. I was thinking we'd be doing it for every file which might be 100/s but I think that is incorrect.